File: hackerrank/counting_valleys.rb
def countingValleys(n, s)
  no_of_valleys = 0
  current_sea_level = 0
  prev_sea_level = current_sea_level

  s.split(//).each do |step|
    if step == 'U'
      prev_sea_level = current_sea_level
      current_sea_level += 1

      if prev_sea_level.negative? && current_sea_level.zero?
        no_of_valleys += 1
      end
    elsif step == 'D'
      prev_sea_level = current_sea_level
      current_sea_level -= 1
      if prev_sea_level.negative? && current_sea_level.zero?
        no_of_valleys += 1
      end
    end
  end

  no_of_valleys
end


n = 8
steps = 'UDDDUDUU'

p countingValleys(n, steps)

File: hackerrank/counting_valleys.go
package main

import "fmt"

func main()  {
	n := 8
	s := "UDDDUDUU"

	fmt.Println(countValleys(n, s))
}

func countValleys(n int, s string) int  {
	var noOfValleys int
	var currentSeaLevel int
	var previousSeaLevel int


	for i := 0; i <= n - 1; i++ {
		step := string(s[i])
		previousSeaLevel = currentSeaLevel

		if step == "U" {
			currentSeaLevel ++
		}

		if step == "D" {
			currentSeaLevel --
		}

		if previousSeaLevel < 0 && currentSeaLevel == 0 {
			noOfValleys ++
		}
	}

	return noOfValleys
}